detectorSegmentations/src/GridDRcalo_k4geo.cpp

```diff
@@ -0,0 +1,232 @@
+}
+
+GridDRcalo_k4geo::GridDRcalo_k4geo(const BitFieldCoder* decoder) : Segmentation(decoder) {
+  // define type and description
+  _type = "GridDRcalo_k4geo";
+  _description = "DRcalo segmentation based on the tower / (Cerenkov or Scintillation) fiber / SiPM hierarchy";
+
+  // register all necessary parameters
+  registerIdentifier("identifier_eta", "Cell ID identifier for numEta", fNumEtaId, "eta");
+  registerIdentifier("identifier_phi", "Cell ID identifier for numPhi", fNumPhiId, "phi");
+  registerIdentifier("identifier_x", "Cell ID identifier for x", fXId, "x");
+  registerIdentifier("identifier_y", "Cell ID identifier for y", fYId, "y");
+  registerIdentifier("identifier_IsCerenkov", "Cell ID identifier for IsCerenkov", fIsCerenkovId, "c");
+  registerIdentifier("identifier_module", "Cell ID identifier for module", fModule, "module");
+
+  fParamBarrel = new DRparamBarrel_k4geo();
+  fParamEndcap = new DRparamEndcap_k4geo();
+}
+
+GridDRcalo_k4geo::~GridDRcalo_k4geo() {
+  if (fParamBarrel) delete fParamBarrel;
+  if (fParamEndcap) delete fParamEndcap;
+}
+
+Vector3D GridDRcalo_k4geo::position(const CellID& cID) const {
+  int noEta = numEta(cID);
+  int noPhi = numPhi(cID);
+
+  DRparamBase_k4geo* paramBase = setParamBase(noEta);
+
+  auto transformA = paramBase->GetSipmTransform3D(noPhi);
+  dd4hep::Position localPos = dd4hep::Position(0.,0.,0.);
+  if ( IsSiPM(cID) ) localPos = dd4hep::Position( localPosition(cID) );
+
+  dd4hep::RotationZYX rot = dd4hep::RotationZYX(M_PI, 0., 0.); // AdHoc rotation, potentially bug
+  dd4hep::Transform3D transformB = dd4hep::Transform3D(rot,localPos);
+  auto total = transformA*transformB;
+
+  dd4hep::Position translation = dd4hep::Position(0.,0.,0.);
+  total.GetTranslation(translation);
+
+  return Vector3D(translation.x(),translation.y(),translation.z());
+}
+
+Vector3D GridDRcalo_k4geo::localPosition(const CellID& cID) const {
+  int numx = numX(cID);
+  int numy = numY(cID);
+  int x_ = x(cID);
+  int y_ = y(cID);
+
+  return localPosition(numx,numy,x_,y_);
+}
+
+Vector3D GridDRcalo_k4geo::localPosition(int numx, int numy, int x_, int y_) const {
+  float ptX = -fGridSize*static_cast<float>(numx/2) + static_cast<float>(x_)*fGridSize + ( numx%2==0 ? fGridSize/2. : 0. );
+  float ptY = -fGridSize*static_cast<float>(numy/2) + static_cast<float>(y_)*fGridSize + ( numy%2==0 ? fGridSize/2. : 0. );
+
+  return Vector3D(ptX,ptY,0.);
+}
+
+/// determine the cell ID based on the position
+CellID GridDRcalo_k4geo::cellID(const Vector3D& localPosition, const Vector3D& /*globalPosition*/, const VolumeID& vID) const {
+  int numx = numX(vID);
+  int numy = numY(vID);
+
+  auto localX = localPosition.x();
+  auto localY = localPosition.y();
+
+  int x = std::floor( ( localX + ( numx%2==0 ? 0. : fGridSize/2. ) ) / fGridSize ) + numx/2;
+  int y = std::floor( ( localY + ( numy%2==0 ? 0. : fGridSize/2. ) ) / fGridSize ) + numy/2;
+
+  return setCellID( numEta(vID), numPhi(vID), x, y );
```

Comment:
since we now copy LHS from RHS, `numEta(vID)` will always return positive `numEta`. To discriminate LHS towers from RHS towers, we can make use of `globalPosition` so that if `globalPosition.z() >= 0` then return `numEta(vID)` and if it is negative then return `-numEta(vID)-1`